docs/src/design/proposals/Registration.md

```diff
@@ -0,0 +1,518 @@
+  zero or more instances of objects of type `X` which have been registered,
+  using a `compositor` function provided as a constructor argument (or via
+  a subsequent `compose` call.)
+  * `register(factory : (function () : X), [options] : RegistrationOptions)`
+    registers an object (as returned by the provided `factory` function)
+    with this provider. Evaluation of the provided `factory` will be
+    deferred until the first `get()` call to the `Provider`.
+  * `compose(compositor : (function (X[]) : S), [options] : RegistrationOptions)`
+    introduces a new strategy for converting an array of registered objects
+    of type `X` into a single instance of an object of type `S`. The
+    highest-priority `compositor` that has been registered in this fashion
+    will be used to assemble the provided object (before decoration)
+* `ServiceProvider<S>` extends `CompositeProvider<S, S>` and provides analogous support
+  for the _composite services_
+  pattern used throughout Open MCT (which, in turn, is a superset of the
+  functionality needed for plain services.)
+* `Registry<T>` extends `CompositeProvider<T, T[]>` and provides analogous support for
+  _extension categories_, also
+  used ubiquitously through Open MCT.
+* `RegistrationOptions` provides a location for additional
+  registration-related metadata to be provided.
+  * `priority` is used to control the ordering and/or preference of
+    factories, compositors, and/or decorators,
+    [as in the existing Framework Layer.](http://nasa.github.io/openmctweb/guide/#priority)
+
+# Examples
+
+The following examples are provided to illustrate the intended usage of
+the Registration API. Particular attention is given to obeying and
+utilizing the "dependency injection as code style"
+[decision from the API Redesign](APIRedesign.md#decisions).
+
+## Building Applications on Open MCT
+
+Applications built using Open MCT are expected to extend the `MCT` base
+class and should typically self-install distinguishing plugins during
+the constructor call. Any pre-installed plugins that are undesirable
+should also be uninstalled at this point (to support such usage,
+`MCT` should expose instances of any installed plugins that are
+considered optional.)
+
+For example:
+
+```
+define(['mct', './plugins'], function (mct, plugins) {
+  var MCT = mct.MCT;
+
+  function Variant() {
+    MCT.call(this);
+
+    this.install(new plugins.SomePlugin());
+    this.install(new plugins.SomeOtherPlugin(this.core));
+
+    this.uninstall(this.plugins.persistence.localStorage);
+  }
+
+  Variant.prototype = Object.create(MCT.prototype);
+
+  return Variant;
+});
+```
+
+Running an application build using Open MCT then typically looks like:
+
+```
+define(['./Variant], function (Variant) {
+  new Variant().run();
+});
+```
+
+## Writing Plugins
+
+A plugin for Open MCT should inherit from the `Plugin` base class.
+
+Plugins will typically use extension points exposed by other plugins;
```

Comment:
These are sharp but completely valid criticisms. _Mea culpa_. There are a few points I disagree with but most are not relevant; the one that sticks out is...

> Throw away bundles. Throw away our current generic extension points. Build a core application which exposes specific APIs. Define extension points where we know we need them. No weird descriptive metadata-- just javascript. 

>  Suddenly, our architecture becomes clear, clean, and easy to explain. It's simple to describe what we do, what we provide. Such things are hard to do when our most truthful description of the platform was as a "plugin extension mechanism". 

... "suddenly."

I love a greenfield as much as anybody. They have none of the problems of legacy systems - they have no problems at all! Unfortunately, they also have none of the solutions of legacy systems. Once you've built those solutions, that "no problems" thing is no longer guaranteed. Maybe you have the same problems; maybe you introduce new ones.

Your approach sounds fine, even attractive, but it's a few sentences that will take a lot of effort to realize, and since it is so different, it will be some time before we fully understand any new problems it introduces. I see high effort and high risk, with possibly high reward (and possibly not.)

The alternative I see is to directly address the problems we have in our current architecture & API, and leave the things that aren't directly problematic as they are. This is moderate effort, low risk,  moderate reward. This is what keeps drawing me back to the status quo, at least for the things which don't appear to be harmful.

I'd like to find a middle ground that addresses at least the most important among your concerns, while still giving us a practical path forward. I don't think we're going to reach consensus on this at the high-level (although I suspect that what we really have in mind is more similar than we are giving each other credit for) so maybe we should resume the conversation when we have something with more specificity to make deltas upon.